File: AoC-2020/advent4/advent4.c
#include <stdio.h>

#define  MAX  1000

struct pass {
     int byr;
     int iyr;
     int eyr;
     char *hgt;
     char *hcl;
     char *ecl;
     int pid;
     int cid;
} passport[MAX];

int main(){
     int i;
     char *field[100];
     char efile;

     for (i = 0; i < 1; ){
          while ( (scanf("%s: ", field) != "\n")){
               printf("%s", field);
               if (field == "byr"){
                    scanf("%d", passport[i].byr);
               printf("%d", passport[i].byr);}
               printf("\n");
               else if (field == "iyr")
                    scanf("%d", passport[i].iyr);
               else if (field == "eyr")
                    scanf("%d", passport[i].eyr);
               else if (field == "hgt")
                    scanf("%s", passport[i].hgt);
               else if (field == "hcl")
                    scanf("%s", passport[i].hcl);
               else if (field == "ecl")
                    scanf("%s", passport[i].ecl);
               else if (field == "pid")
                    scanf("%d", passport[i].pid);
               else if (field == "cid")
                    scanf("%d", passport[i].cid);
          }
          i++;
     }
}

File: AoC-2020/advent4/advent4.py
import re

def check_pass(file):

    pass_list = file.read().split('\n\n')

    fields = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
    valid_number = 0

    for i in range(len(pass_list)):
        if all(x in pass_list[i] for x in fields):
            valid_number += 1
            byr = int(re.search("(?<=byr: )\d\d\d\d", pass_list[i]).group(0))
            iyr = int(re.search("(?<=iyr: )\d\d\d\d", pass_list[i]).group(0))
            eyr = int(re.search("(?<=eyr: )\d\d\d\d", pass_list[i]).group(0))
            hgt = re.search("(?<=hgt: ).*[cm|in]", pass_list[i])
            while hgt:
                try:
                    hgt = str(hgt.group(0))
                except AttributeError:
                    print(hgt)
                    break
            if hgt and "cm" in hgt:
                try:
                    hgtcm = int(re.search(".*(?=cm|in)", hgt).group(0))
                except:
                    hgtcm = 0
            else:
                hgtcm = 0
            if hgt and "in" in hgt:
                hgtin = int(re.search(".*(?=cm|in)", hgt).group(0))
            else:
                hgtin = 0

            if byr >= 1920 and byr <= 2002 and \
            iyr >= 2010 and iyr <= 2020 and \
            eyr >= 2020 and eyr <= 2030 and \
            ((hgtcm >= 150 and hgtcm <= 193) or (hgtin >= 59 and hgtin <= 76)):

                print(True, hgtcm, hgtin)
            else:
                print(False, hgtcm, hgtin)

    print(len(pass_list))
    print(valid_number)

file = open('input')
check_pass(file)
